src/wallet.rs

```diff
@@ -519,4 +540,50 @@ mod tests {
         let to_keypair = wallet.decrypt(password).expect("wallet to keypair");
         assert_eq!(from_keypair, to_keypair);
     }
+
+    #[test]
+    fn keypair_from_swarm() {
+        use crate::keypair::{KeyType, Network};
+
+        let mut swarm_key: &[u8] = &[
+            0x00, // <- Type/Network: ecc_compact/mainnet
+            // Private exponent
+            0xd3, 0x98, 0xd1, 0xfc, 0x3f, 0x4c, 0x22, 0x79, 0xa8, 0x9a, 0x25, 0xb2, 0xd3, 0x52,
+            0xcf, 0x38, 0x25, 0xcf, 0xc0, 0x18, 0x8d, 0xd7, 0xee, 0xb3, 0x04, 0x7c, 0x29, 0x29,
+            0x0b, 0xdf, 0xaa, 0xd2,
+            // End private exponent
+            // Begin public key
+            0x04, // <- ECC Point tag: Full X/Y pair
+            // Begin X-coordinate
+            0x33, 0x4b, 0xb9, 0x5a, 0x68, 0x15, 0x4a, 0x74, 0xed, 0xf6, 0x51, 0x81, 0x6b, 0xd5,
+            0x56, 0x11, 0xb5, 0xba, 0x85, 0x04, 0x3f, 0xbd, 0x20, 0xa3, 0x74, 0xa3, 0x83, 0x1b,
+            0x83, 0x98, 0x04, 0x35,
+            // End X-coordinate
+            // Begin Y-coordinate
+            0x2e, 0x7d, 0xcd, 0x51, 0xa1, 0x90, 0x92, 0x63, 0x77, 0x9e, 0x4d, 0x0c, 0x6d, 0xb5,
+            0x2e, 0x23, 0x2c, 0x9b, 0xdb, 0x64, 0xe1, 0x94, 0x24, 0x08, 0x99, 0x1c, 0x4b, 0x29,
+            0xac, 0xd7, 0x6c, 0xce,
+            // End Y-coordinate
+            // End public key
+        ];
+        let from_swarm = Keypair::read(&mut swarm_key).unwrap();
+
+        let seed_words: Vec<String> = [
+            "squeeze", "shoot", "lecture", "leader", "season", "devote", "pen", "dwarf", "ready",
+            "once", "record", "icon", "friend", "theme", "giraffe", "road", "upgrade", "oblige",
+            "business", "false", "mouse", "used", "prize", "foster",
+        ]
+        .iter()
+        .map(|&s| s.to_owned())
+        .collect();
+
+        let tag = KeyTag {
+            key_type: KeyType::EccCompact,
+            network: Network::MainNet,
+        };
+
+        let from_seed = gen_keypair(tag, Some(seed_words), Some(&SeedType::Bip39)).unwrap();
+
+        assert_eq!(from_seed, from_swarm);
```

Comment:
Can you add a test for an ed25519 swarm key as well? Many validators use one of them

Reply:
Done. I've also diagonalized the test matrix by making it a testnet test, as well.